lib/sycamore/sycamore/writers/base.py

```diff
@@ -0,0 +1,160 @@
+from dataclasses import dataclass
+from typing import Any, Iterable, Optional
+
+from abc import ABC, abstractmethod
+import numpy as np
+
+from ray.data import Dataset, Datasink
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.execution.interfaces import TaskContext
+from ray.data.block import Block, BlockAccessor
+from sycamore.data.document import Document, MetadataDocument
+from sycamore.plan_nodes import Node, Write
+
+
+class BaseDBWriter(Write):
+
+    # Type param for the client
+    class client_t(ABC):
+        @classmethod
+        @abstractmethod
+        def from_client_params(cls, params: "BaseDBWriter.client_params_t") -> "BaseDBWriter.client_t":
+            pass
+
+        @abstractmethod
+        def write_many_records(self, records: list["BaseDBWriter.record_t"]):
+            pass
+
+        @abstractmethod
+        def create_index_if_missing(self, index_params: "BaseDBWriter.index_params_t"):
+            pass
+
+    # Type param for the objects to write to the db
+    class record_t(ABC):
+
+        @classmethod
+        @abstractmethod
+        def from_doc(cls, document: Document) -> "BaseDBWriter.record_t":
+            pass
+
+        @abstractmethod
+        def serialize(self) -> bytes:
+            pass
+
+        @classmethod
+        @abstractmethod
+        def deserialize(cls, byteses: bytes) -> "BaseDBWriter.record_t":
+            pass
+
+    # Type param for the object used to configure a new index if necessary
+    @dataclass
```

Comment:
This works for OpenSearch and Pinecone which both refer to indexes. Weaviate seems to use collection. SQL uses table, S3 uses bucket, filesystems would use directory. Maybe DestinationParams since users will be specifiying a destination for the writes?

Reply:
Going with TargetParams. it's shorter